Add connection related functionality to VisualShader (deleting, dropping/inserting a node)

Geometror opened the pull request:
![VS Connections](https://github.com/godotengine/godot/assets/50084500/2be25d06-b947-4204-a945-e1cbfbff162a)
To demonstrate the new connection functionality, I went ahead and implemented some UX improvements for VS.

Depends on #86158 [merged] ~~#83508~~.
- Add a context menu for connections in the VisualShader editor consisting of
	- Disconnect: Deletes the connection
	- Insert New Node: Opens the "Create Shader Node" dialog and inserts the chosen node by splitting the original connection.

TODO:
- [x] Allow dragging fully unconnected nodes on connections to insert them
- [x] Polish

YuriSizov commented:
Amazing! Do you think adding a knife tool would be hard to implement here?

Geometror commented:
Small update: Unconnected nodes can now be dropped on connections!
![VS Connections DROP](https://github.com/godotengine/godot/assets/50084500/35f43565-6691-40b8-89ef-a6f894cb1720)

QbieShay commented:
OMG YES!

AThousandShips commented on scene/gui/graph_edit.h:
```diff
@@ -112,12 +112,25 @@ class GraphEdit : public Control {
 
 public:
-	struct Connection {
+	struct Connection : RefCounted {
```
Is it safe to use `RefCounted` like this without declaring the class with `GDCLASS`? Or registering it?

Geometror replied:
Currently it seems like the only option since we don't allow std::shared_ptr. I'm not a huge fan of this solution either since RefCounted is quite heavy ~352 Bytes and is indeed meant for types which are registered (but since its just an internal struct I don't want to register it). I'd prefer having an equivalent to `std::shared_ptr` for this, but that's out of the scope of this PR.

AThousandShips replied:
Understand, mainly asked for the actual functionality, if the type isn't registered it isn't guaranteed to work correctly with recounting

Geometror replied:
Just for reference: We discussed this on RocketChat. 
This should be fine since there are other classes/structs which aren't registered and derive from `RefCounted`. You can find some of them using the regex `: public RefCounted.*\n(\t)*^((?!GDCLASS).)*$` (18 results).

AThousandShips commented on editor/plugins/visual_shader_editor_plugin.cpp:
```diff
@@ -3561,6 +3576,132 @@ void VisualShaderEditor::_connection_from_empty(const String &p_to, int p_to_slo
+	// Check whether the dragged node was dropper over a connection.
+	List<Ref<GraphEdit::Connection>> intersecting_connections = graph->get_connections_intersecting_with_rect(selected_node_rect);
+
+	if (intersecting_connections.size() == 0) {
```
```suggestion
	if (intersecting_connections.is_empty()) {
```
Preferred syntax for readability

AThousandShips reviewed (approved):
LGTM

YuriSizov commented:
Needs a rebase

Geometror commented:
Rebased, but this PR depends on #86158. That one needs to be merged first, this PR is just already based on the connections rework branch.

YuriSizov commented:
Okay, I'll mark it as a draft for the time being then.

Geometror commented:
Rebased after #86158 got merged.
Tested and it looks like all features still work fine.

akien-mga commented on editor/plugins/visual_shader_editor_plugin.cpp:
```diff
@@ -3191,6 +3191,15 @@ void VisualShaderEditor::_add_node(int p_idx, const Vector<Variant> &p_ops, Stri
 
 	bool created_expression_port = false;
 
+	// A node is inserted in a already present connection.
```
```suggestion
	// A node is inserted in an already present connection.
```

akien-mga commented on editor/plugins/visual_shader_editor_plugin.cpp:
```diff
@@ -3564,6 +3579,132 @@ void VisualShaderEditor::_connection_from_empty(const String &p_to, int p_to_slo
+		return false;
+	}
+
+	// Check whether the dragged node was dropper over a connection.
```
```suggestion
	// Check whether the dragged node was dropped over a connection.
```

akien-mga commented on editor/plugins/visual_shader_editor_plugin.cpp:
```diff
@@ -3564,6 +3579,132 @@ void VisualShaderEditor::_connection_from_empty(const String &p_to, int p_to_slo
+	EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
+	undo_redo->create_action(TTR("Insert node"));
+
+	// Check whether the dragged node was dropper over a connection.
```
```suggestion
	// Check whether the dragged node was dropped over a connection.
```

akien-mga commented on editor/plugins/visual_shader_editor_plugin.cpp:
```diff
@@ -5238,6 +5449,7 @@ VisualShaderEditor::VisualShaderEditor() {
 	graph->connect("gui_input", callable_mp(this, &VisualShaderEditor::_graph_gui_input));
 	graph->connect("connection_to_empty", callable_mp(this, &VisualShaderEditor::_connection_to_empty));
 	graph->connect("connection_from_empty", callable_mp(this, &VisualShaderEditor::_connection_from_empty));
+	// graph->connect("end_node_move", callable_mp(this, &VisualShaderEditor::_handle_node_drop_on_connection));
```
Dead code?

Geometror replied:
Yes.

Geometror committed:
Add connection-related VisualShader operations
- Insert a node in a connection
- Drop an unconnected node on a connection to insert it
- Delete a connection

YuriSizov commented:
Thanks! If there are further changes required in the VS plugin, we can do them in a follow-up.

(And I personally still want this knife tool :P )
